### DISCUSSION TITLE
Geometry / Compute shaders?

### DISCUSSION THREAD

**Discussion by thedocruby** [FyroxEngine/Fyrox#368] [Q&A]
I really like what Fyrox has done with their modern graphics and rendering pipeline, but I can't seem to find anything about how to add Geometry or Compute programs to shaders. So, I have a few questions:

- Are Geometry and/or Compute shaders a feature of FyroxEngine? (they should be)
- If so, where can I find documentation and/or examples?
- If not, are there plans to add Geometry or Compute shader support in the future? What can I do to help accelerate the development process in that area?

Like I said, really a great graphics pipeline, and in general probably the best all-around modern rust game engine so far, but without support for Geometry and Compute shaders, I might need to look to more graphics-centered engines and frameworks to support my project (which means I will have to spend more time setting up things like the scene and lighting by hand, and those things really aren't the focus of my project. If I can get geometry & compute shaders working with FyroxEngine, I'd rather use Fyrox for it's easy GUI, so I can get the basic stuff over with quickly & painlessly and have more time for the meat of the project, a proof-of-concept for my fancy new render technique)

I appreciate any help or advice you have for me.

**Comment by mrDIMAS** [comment]
Hi! Geometry shaders are not available because I didn't use them in the renderer, such shaders could be added to the [framework](https://github.com/FyroxEngine/Fyrox/tree/master/src/renderer/framework) pretty easily. You can even use raw OpenGL (not quite actually, the engine uses `glow` crate, but its almost the same) functions for this in your custom render pass. There is full access to low level details via `Renderer::state.gl`, the book has a separate chapter of [how to create custom render passes](https://fyrox-book.github.io/fyrox/rendering/render_pass.html).

Compute shaders are not available because of OpenGL and macOS support (and also because I didn't use them). macOS only implements OpenGL 4.1 which does not provide access to compute shaders, because they're only available in 4.3 Core Profile. It is possible to add it via optional functionality, but I just didn't bother with it. 

The renderer is not very advanced, the last time I touched the renderer is more than a year already. Currently focusing on other more important things, graphics could be improved in a few months anyways. At some day I'll also implement more modern renderer using `wgpu`, but that's for the future.
